### ISSUE TITLE
Review decision to require `RootJsonFormat` to marshal from/to Json

### ISSUE THREAD

**Issue by jrudolph** [akka/akka#17775]
The latest [JSON RFC 7159](https://tools.ietf.org/html/rfc7159) is less strict than other specifications before:

> A JSON text is a serialized value.  Note that certain previous
> specifications of JSON constrained a JSON text to be an object or an
> array.  Implementations that generate only objects or arrays where a
> JSON text is called for will be interoperable in the sense that all
> implementations will accept these as conforming JSON texts.

In spray-json the former restriction was modelled by `RootJsonFormat` which only consumes or produces json arrays or objects. Removing the restriction would make using spray-json with akka-http simpler with the potential disadvantage that we encourage a format that some peers possibly cannot understand.

On the other hand, changing to requiring `JsonFormat` would mean that suddenly Marshallers to/from most primitive types would be in scope that are likely to interfere with other ones.

/cc @sirthias

**Comment by jrudolph** [comment]
See also this recent akka-user discussion: https://groups.google.com/d/topic/akka-user/hicTjKXHGeg/discussion

**Comment by rkuhn** [comment]
What about adding an implicit that needs to be imported separately that allows the marshalling of JsonFormat?

**Comment by sirthias** [comment]
Yeah, I like the idea of making things configurable with another implict, which you'd have to bring into scope in order to enable the marshalling of non-root JsonFormats.

**Comment by fommil** [comment]
Cool, just seen this. I have an alternative `SprayJsonSupport` object that I bring into scope as an alternative when an external API demands non-root objects.

I just wanted to bring something to your attention regarding derived formats that you might not have thought of.

The compiler ignores other container types when doing implicit resolution, so if the user provides a `RootJsonFormat[T]` **and** a `JsonFormat[T]` the compiler treats them as completely independent (and will not flag an ambiguous implicit). If only `RootJsonFormat[T]` is provided, then it will be searched only after the compiler confirms that there is no `JsonFormat[T]`.

When using automatic derivers, e.g. [`spray-json-shapeless`](https://github.com/fommil/spray-json-shapeless), `RootJsonFormat[T]` will be created automatically, but the user could provide a `JsonFormat[T]` which would silently win this battle **only when the `JsonFormat[T]`** is requested. This might confuse some users, because previously the automatically derived format will have won.

We have a testcase for this exact point, but it's currently disabled because it doesn't compile under 2.10 https://github.com/fommil/spray-json-shapeless/blob/master/src/test/scala/fommil/sjs/FamilyFormatsSpec.scala#L238

**Comment by ktoso** [comment]
Marking as triaged, we'll want to do some small improvements to the marshalling infra anyway.

**Comment by ktoso** [comment]
Ticket moved to github.com/akka/akka-http  (direct link to new issue visible on github just above this comment). Rationale for the move discussed here: akka/akka-meta#27

If you are interested in contributing or tracking this issue, please comment in akka-http instead from now on :-)
